tree/ntuple/v7/src/RField.cxx

```diff
@@ -916,6 +923,181 @@ void ROOT::Experimental::RClassField::AcceptVisitor(Detail::RFieldVisitor &visit
+      case EDataType::kLong_t:
+      case EDataType::kLong64_t:
+         itemField = std::make_unique<RField<std::int64_t>>("_0");
+         break;
+      case EDataType::kULong_t:
+      case EDataType::kULong64_t:
+         itemField = std::make_unique<RField<std::uint64_t>>("_0");
+         break;
+      case EDataType::kFloat_t:  itemField = std::make_unique<RField<float>>("_0"); break;
+      case EDataType::kDouble_t: itemField = std::make_unique<RField<double>>("_0"); break;
+      case EDataType::kBool_t:   itemField = std::make_unique<RField<bool>>("_0"); break;
+      default:
+         throw RException(R__FAIL("unsupported value type"));
+      }
+   }
+   fItemSize = itemField->GetValueSize();
+   Attach(std::move(itemField));
+}
+
+std::unique_ptr<ROOT::Experimental::Detail::RFieldBase>
+ROOT::Experimental::RCollectionClassField::CloneImpl(std::string_view newName) const
+{
+   return std::unique_ptr<RCollectionClassField>(
+      new RCollectionClassField(newName, GetType(), fProxy->GetCollectionClass()));
+}
+
+std::size_t ROOT::Experimental::RCollectionClassField::AppendImpl(const Detail::RFieldValue &value)
+{
+   TVirtualCollectionProxy::TPushPop RAII(fProxy, value.GetRawPtr());
+   std::size_t nbytes = 0;
+   auto count = fProxy->Size();
+   for (unsigned i = 0; i < count; ++i) {
+      auto itemValue = fSubFields[0]->CaptureValue(fProxy->At(i));
+      nbytes += fSubFields[0]->Append(itemValue);
+   }
+   Detail::RColumnElement<ClusterSize_t> elemIndex(&fNWritten);
+   fNWritten += count;
+   fColumns[0]->Append(elemIndex);
+   return nbytes + sizeof(elemIndex);
+}
+
+void ROOT::Experimental::RCollectionClassField::ReadGlobalImpl(NTupleSize_t globalIndex, Detail::RFieldValue *value)
+{
+   TVirtualCollectionProxy::TPushPop RAII(fProxy, value->GetRawPtr());
+
+   ClusterSize_t nItems;
+   RClusterIndex collectionStart;
+   fPrincipalColumn->GetCollectionInfo(globalIndex, &collectionStart, &nItems);
+
+   auto oldNItems = fProxy->Size();
+   if (fProxy->GetProperties() & TVirtualCollectionProxy::kNeedDelete) {
+      for (std::size_t i = 0; i < oldNItems; ++i) {
+         auto itemValue = fSubFields[0]->CaptureValue(fProxy->At(i));
+         fSubFields[0]->DestroyValue(itemValue, true /* dtorOnly */);
+      }
+   }
+   fProxy->Clear();
+
+   // Avoid heap fragmentation at the cost of temporarily allocating slightly more memory
+   const size_t buffSize = std::max(kReadChunkSize, fItemSize);
+   const std::uint32_t nItemsPerChunk = buffSize / fItemSize;
+   auto buff = std::make_unique<unsigned char[]>(buffSize);
+
+   auto nItemsLeft = static_cast<std::uint32_t>(nItems);
+   while (nItemsLeft > 0) {
+      auto count = std::min(nItemsLeft, nItemsPerChunk);
+      for (std::size_t i = 0; i < count; ++i) {
+         auto itemValue = fSubFields[0]->GenerateValue(buff.get() + (i * fItemSize));
+         fSubFields[0]->Read(collectionStart + i, &itemValue);
+      }
+      fProxy->Insert(buff.get(), value->GetRawPtr(), count);
```

Comment:
Another optimization opportunity:  Rather than calling `Insert`, `TBranchElement/TStreamerInfo` does:
* resize/create the element
* iterate over the elements and update in place.

Reply:
But that only applies for sequential containers, e.g. `kSTLvector`, no?  I believe other containers might still need element-by-element insertion.

Reply:
The pattern can be applied to all containers.  Even with non-vector containers it is an optimization as it cut the number of allocation (of user objects) in half (i.e only the final object rather than create a temporary object and copy over to the final object).